test/amqp/high_level_test.rb

```diff
@@ -134,4 +134,450 @@ def test_it_can_resubscribe_on_reconnect
       client.stop
     end
   end
+
+  def test_fanout_convenience_method
+    msgs = Queue.new
+    client = AMQP::Client.new("amqp://#{TEST_AMQP_HOST}").start
+    begin
+      e = client.fanout("test.fanout", auto_delete: true)
+      q = client.queue("test.fanout.queue")
+      q.bind("test.fanout", "")
+      q.subscribe do |msg|
+        msgs.push msg
+      end
+      e.publish("fanout message", "")
+
+      msg = msgs.pop
+      assert_equal "fanout message", msg.body
+      assert_equal "test.fanout", msg.exchange_name
+
+      e.delete
+      q.delete
+    ensure
+      client.stop
+    end
+  end
+
+  def test_direct_convenience_method
+    msgs = Queue.new
+    client = AMQP::Client.new("amqp://#{TEST_AMQP_HOST}").start
+    begin
+      e = client.direct("test.direct", auto_delete: true)
+      q = client.queue("test.direct.queue")
+      q.bind("test.direct", "direct.key")
+      q.subscribe do |msg|
+        msgs.push msg
+      end
+      e.publish("direct message", "direct.key")
+
+      msg = msgs.pop
+      assert_equal "direct message", msg.body
+      assert_equal "test.direct", msg.exchange_name
+      assert_equal "direct.key", msg.routing_key
+
+      e.delete
+      q.delete
+    ensure
+      client.stop
+    end
+  end
+
+  def test_topic_convenience_method
+    msgs = Queue.new
+    client = AMQP::Client.new("amqp://#{TEST_AMQP_HOST}").start
+    begin
+      e = client.topic("test.topic", auto_delete: true)
+      q = client.queue("test.topic.queue")
+      q.bind("test.topic", "user.*")
+      q.subscribe do |msg|
+        msgs.push msg
+      end
+      e.publish("topic message", "user.created")
+
+      msg = msgs.pop
+      assert_equal "topic message", msg.body
+      assert_equal "test.topic", msg.exchange_name
+      assert_equal "user.created", msg.routing_key
+
+      e.delete
+      q.delete
+    ensure
+      client.stop
+    end
+  end
+
+  def test_headers_convenience_method
+    msgs = Queue.new
+    client = AMQP::Client.new("amqp://#{TEST_AMQP_HOST}").start
+    begin
+      e = client.headers("test.headers", auto_delete: true)
+      q = client.queue("test.headers.queue")
+      q.bind("test.headers", "", arguments: { type: "order", "x-match": "all" })
+      q.subscribe do |msg|
+        msgs.push msg
+      end
+      e.publish("headers message", "", headers: { type: "order", priority: "high" })
+
+      msg = msgs.pop
+      assert_equal "headers message", msg.body
+      assert_equal "test.headers", msg.exchange_name
+      assert_equal({ "type" => "order", "priority" => "high" }, msg.properties.headers)
+
+      e.delete
+      q.delete
+    ensure
+      client.stop
+    end
+  end
+
+  def test_convenience_methods_equivalent_to_exchange_method
+    client = AMQP::Client.new("amqp://#{TEST_AMQP_HOST}").start
+    begin
```

Comment:
Probably easier to debug if it's one set of assertions per type in separate test cases. but I'll allow it:)